src/macaron/slsa_analyzer/checks/two_person_reviewed_check.py

```diff
@@ -23,37 +20,21 @@
 logger: logging.Logger = logging.getLogger(__name__)
 
 
-class TwoPersonReviewedTable(CheckFacts, ORMBase):
```

Comment:
Why did you remove the table? This table is necessary for users to write policies and validate the check results using the policy engine.

Reply:
All right, I will restore it.
I misunderstood that this is used to store results.

Reply:
It is used to persist the results in a database, which is used by the policy engine separately.

Reply:
I decided to remove it since I also checked the vsc_check.py and there is not table defined. Does this mean the results from the VCS check will not be consumed by the policy engine?

Reply:
That's correct. We should add a table for the VCS check too (not added yet because we didn't have a use case so far). But we need to make the check result available to the policy engine in your case.

Reply:
`[CRITICAL] Database error (sqlite3.IntegrityError) NOT NULL constraint failed: _analysis.repository
[SQL: INSERT INTO _analysis (analysis_time, macaron_version) VALUES (?, ?)]
[parameters: ('2023-09-01T04:47:53+00:00', '0.3.0')]
(Background on this error at: https://sqlalche.me/e/20/gkpj)`

**Local integration test fail due to this error. Does this refer to primary key or nullable issue?
This error cause the issue of "unable to generate report"**.

Reply:
Hi. Could you let us know which case in the integration tests is failing? You should look for the line
```FAILED integration test ...```
in the log.

Reply:
`Error: FAILED integration test (line 37)
Error: FAILED integration test (line 44)`
Some other similar errors results from analysis function.

Reply:
In Macaron 0.3.0 we changed the data model and the DB constraints have changed. You should make sure to remove any existing `macaron.db` file from your output directory to avoid this error.

Reply:
Hi there. The check is implemented through multi-threading method in order to tackle with io-bound issue. Since the check requires scanning through numerous PRs, GitHub API rate limit mechanism will force the system to sleep for a long time. 
Due to the this issue, the integration test will be canceled. 
Also, numerous API request causes unstable results. The expected results might not be generated, which results in integration test fail.  
Would it be possible to handle rate limit and fetch the data in a stable way?

Reply:
Given this issue, I guess that we are trying to check too many PRs, which results in in-deterministic results. Is there a way that we only limit the range of the opened/closed PRs so that we don't have to perform too many queries?

> The check is implemented through multi-threading method in order to tackle with io-bound issue.

I don't think it's possible to implement the checks using multi-threading as we haven't considered multi-threading inside checks for our current infrastructure.

In additions, I also have some questions on the logic of your proposed implementation:

1. Fetching all PRs number, and it is able to specify target branch for - Does this mean that we will check for **all** PRs existed inside a repository?  
2. Checking commit time is existed, and this step will preserve the merge and closed PRs - What do you mean by `commit time` here and what is the meaning of `preserve` the merge and closed PRs.

> Would it be possible to handle rate limit and fetch the data in a stable way?

Given the nature of GitHub API, the current solution is to check if we exceed the GitHub API limit and wait until the rate limit is removed (implementation [here](https://github.com/oracle/macaron/blob/0f426982837708539fb64b76001b823b79596b0e/src/macaron/util.py#L103)). However, this implementation is not ideal too. 
I guess in terms of your check, we might need to think of a better heuristic to query only subset of PRs to reduce the number of queries. (cc @behnazh-w) 
In terms of user experience, taking too much time on a single check wouldn't be ideal after all.

Reply:
> Given this issue, I guess that we are trying to check too many PRs, which results in in-deterministic results. Is there a way that we only limit the range of the opened/closed PRs so that we don't have to perform too many queries?
I will try moving from REST to GraphQL, since it will allow fetching the precise content we need. Currently, I have specified the parameter to 'closed' only.  
> I don't think it's possible to implement the checks using multi-threading as we haven't considered multi-threading inside checks for our current infrastructure. 
Alright, I'll modify my design.
> 1. Fetching all PRs number, and it is able to specify target branch for - Does this mean that we will check for all PRs existed inside a repository?
Yep, and now, the scope has been scale down to only the merge PRs.
> 2. Checking commit time is existed, and this step will preserve the merge and closed PRs - What do you mean by commit time here and what is the meaning of preserve the merge and closed PRs.
I think **commit time** should be **merge_at** key word that ensures the PR has been merged. My approach is requesting closed PR information, then check the **merge_at** keyword existed. The closed PR does not guarantee it is being merged.
>  In terms of user experience, taking too much time on a single check wouldn't be ideal after all.
I will figure out and test different approach, then updating the performance of the approach in this thread.